### DISCUSSION TITLE
How can I get the length of BYTE*(unsigned char*) data ? ( bitmap data )

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#7940] [Q&A] [answered]
hello. I am a student making rdp with access control applied using freerdp library.

I am implementing server connection and events in java using JNI.

In order to capture the screen image by time, I am trying to receive the data value from the endpaint part as a JNI callback.

In order to send the data value, BYTE* ( unsigned char*) , to the JNI callback, you need to create a JArray of JNI and put data in it.

Therefore, the size of the array of data is required, but since it is a pointer, the size of the array cannot be measured with sizeof.

I've been looking for a function like _mszie and several methods, but I've spent a lot of time.

How can I get the length of a data array?

I mainly tested on /client/Mac/MRDPView.m - mac_end_paint

I succeeded in outputting the data value to be moved as an image using the stb_image library.

i use 

rdpContext->rdpGdi->primary->bitmap->data

Can you please tell me how to get the data array in this structure.

I put a lot of time into it but failed..

`BOOL mac_end_paint(rdpContext *context)
{

	rdpGdi *gdi;
	HGDI_RGN invalid;
	NSRect newDrawRect;
	int ww, wh, dw, dh,wd;
	mfContext *mfc = (mfContext *)context;
	MRDPView *view = (MRDPView *)mfc->view;
	gdi = context->gdi;

        time_t t;
            clock_t timee = clock();
            /* Intializes random number generator */
            srand((unsigned) time(&timee));
            /* Print 5 random numbers from 0 to 49 */

            int a = rand() % 10000;
               printf("%d\n", a);

            char mystr[5];
           sprintf(mystr , "%d", a);
           printf("%s\n",mystr);
             char locationName[200] = "/Users/kimfk567/Documents/JNAImage/";
             char bmp[5] = ".bmp";
            strncat(locationName,mystr,5);
            strncat(locationName,bmp,5);
            printf("locationName %s\n", locationName);
            stbi_write_bmp(locationName, gdi->primary->bitmap->width, gdi->primary->bitmap->height, 4, data);

    BYTE *s1 = gdi->primary->bitmap->data;
    FILE *fp = fopen("/Users/kimfk567/Documents/JNAText2.txt", "wb");
    fwrite(s1, sizeof(s1), 1, fp);
    fclose(fp);

            int dataSize = sizeof(gdi->primary->bitmap->data);
            JNIEnv* env;
             jni_attach_thread(&env);
            jbyteArray retArr = (*env)->NewByteArray(env,malloc_size(gdi->primary->bitmap->data));
  

           (*env)->SetByteArrayRegion(env, retArr, 0, malloc_size(gdi->primary->bitmap->data) , gdi->primary->bitmap->data);

	freerdp_callback("JNIcallback", "(II[BCCCCCCCC)V",
	              gdi->primary->bitmap->width, gdi->primary->bitmap->height, retArr ,src[0],src[1],src[2],src[3],src[4],src[5],src[6],src[7]);
        free(arr);
                		jni_detach_thread();

        }

	if (!gdi)
		return FALSE;

	ww = mfc->client_width;
	wh = mfc->client_height;
	dw = mfc->common.context.settings->DesktopWidth;
	dh = mfc->common.context.settings->DesktopHeight;

	if ((!context) || (!context->gdi))
		return FALSE;

	if (context->gdi->primary->hdc->hwnd->invalid->null)
		return TRUE;

	invalid = gdi->primary->hdc->hwnd->invalid;
	newDrawRect.origin.x = invalid->x;
	newDrawRect.origin.y = invalid->y;
	newDrawRect.size.width = invalid->w;
	newDrawRect.size.height = invalid->h;

	if (mfc->common.context.settings->SmartSizing && (ww != dw || wh != dh))
	{
		newDrawRect.origin.y = newDrawRect.origin.y * wh / dh - 1;
		newDrawRect.size.height = newDrawRect.size.height * wh / dh + 1;
		newDrawRect.origin.x = newDrawRect.origin.x * ww / dw - 1;
		newDrawRect.size.width = newDrawRect.size.width * ww / dw + 1;
	}
	else
	{
		newDrawRect.origin.y = newDrawRect.origin.y - 1;
		newDrawRect.size.height = newDrawRect.size.height + 1;
		newDrawRect.origin.x = newDrawRect.origin.x - 1;
		newDrawRect.size.width = newDrawRect.size.width + 1;
	}

	windows_to_apple_cords(mfc->view, &newDrawRect);
	dispatch_sync(dispatch_get_main_queue(), ^{
		[view setNeedsDisplayInRect:newDrawRect];
	});
	gdi->primary->hdc->hwnd->ninvalid = 0;
	return TRUE;
}`

**Comment by akallabeth** [comment, marked as answer]
* It is `gdi->stride * gdi->height`.
* `gdi->stride` is the `gdi->width * <bytes per pixel> + <padding per line>` if you do full updates, otherwise you need to check the damage rectangles for offsets/sizes
* The `if (mfc->common.context.settings->SmartSizing && (ww != dw || wh != dh))` is only relevant if you do software scaling (e.g. you local window has different size than the session resolution)

That is if you implement a RDP client, you do that right?

**Reply by kimcoldwater** [reply]
thanks for the kind reply.
If it's not rude, can I ask you one more question?
Do you know which one to use, JByteArray or JCharArray when converting BYTE*( unsigned char*) using JNI?
Currently, we are referencing the Android implementation a lot.
I can confirm the later part through several tests, but I spent too much time.
 It would be of great help to me if you could give me advice on the logic of receiving data from the Freerdp module to JNI and outputting the image in JAVA.

**Reply by akallabeth** [reply]
@kimcoldwater depends on what you want to do with the data later on. Since sooner or later you´ll end up feeding the data to `BitmapFactory.decodeByteArray` or similar most likely `JByteArray`, but you can also directly convert using native Bitmap and pass that via JNI https://developer.android.com/ndk/reference/group/bitmap

**Reply by kimcoldwater** [reply]
@akallabeth 
Thank you for your answer.
I am currently in a development environment where Android studio are not available.
I received the data as java using JByteArray, and I am using  
java.awt.image.BufferedImage 
and javax.imageio.ImageIO
but the image is not output.
Do you know how to print the data value as an image?

 ```
 public static void JNIcallback(int width, int height, byte[] data
    ) throws IOException {
        //image == null
            ByteArrayInputStream inputStream = new ByteArrayInputStream(data);
            BufferedImage bufferedImage = ImageIO.read(inputStream);
            ImageIO.write(bufferedImage, "png", new File("/Users/kimfk567/Documents/javaJNI_Image/data6.png"));
        
        // byte->int
        ByteArrayInputStream inputStream = new ByteArrayInputStream(data);
            BufferedImage bufferedImage = ImageIO.read(inputStream);
            BufferedImage a = new BufferedImage(1024,1024,TYPE_4BYTE_ABGR);
            WritableRaster raster = (WritableRaster) a.getData();
           raster.setPixels(0,0,100,100,[Byte to int array);
           a.setData(raster);            
           ImageIO.write(bufferedImage, "bmp", new File("/Users/kimfk567/Documents/javaJNI_Image/data1.bmp"));
        } 
```

error

```
Exception in thread "Thread-31" java.lang.IllegalArgumentException: image == null!
	at java.desktop/javax.imageio.ImageTypeSpecifier.createFromRenderedImage(ImageTypeSpecifier.java:925)
	at java.desktop/javax.imageio.ImageIO.getWriter(ImageIO.java:1610)
	at java.desktop/javax.imageio.ImageIO.write(ImageIO.java:1542)
	at JNI.helloJ.JNIcallback(helloJ.java:108)

```

**Reply by akallabeth** [reply]
@kimcoldwater better join our matrix channel for such discussions.

**Reply by kimcoldwater** [reply]
@akallabeth 
Thanks to your help, we solved it Thank you very much. 
I'll join the Matrix channel
![image](https://user-images.githubusercontent.com/84696467/171611780-798599d5-a1b8-4658-8735-56d214b5ad8e.png)
